File: backend/app/api/v1/decision_rooms.py
"""
REST API Router for Phase 53 — Unified Human-AI Collaboration, Decision Room & Augmented Intelligence Platform.
"""

from typing import Dict, Any, List, Optional
from fastapi import APIRouter, HTTPException, Query, status

from backend.app.schemas.decision_rooms import (
    DecisionRoomCreateRequest,
    DecisionStatusTransitionRequest,
    RecordDecisionRequest,
    SetContextRequest,
    AddEvidenceRequest,
    AddAssumptionRequest,
    AddUnknownRequest,
    CreateOptionRequest,
    AddCriterionRequest,
    ScoreOptionRequest,
    SubmitAnalysisRequest,
    SubmitAdversarialReviewRequest,
    AddCommentRequest,
    RecordApprovalRequest,
    CreateActionRequest,
    RecordOutcomeRequest,
    SubmitPostReviewRequest,
    CopilotQueryRequest,
)
from backend.app.services.decision_rooms.service import global_decision_room_service
from backend.app.services.decision_rooms.base import (
    DecisionType,
    DecisionImportance,
    DecisionStatus,
    EvidenceType,
    StatementCategory,
    SpecialistRole,
    ApprovalStatus,
)

router = APIRouter(prefix="/decision-rooms", tags=["Decision Rooms & Human-AI Collaboration"])


@router.get("", summary="List Decision Rooms")
async def list_decision_rooms(
    status: Optional[str] = Query(None),
    decision_type: Optional[str] = Query(None),
    importance: Optional[str] = Query(None),
) -> Dict[str, Any]:
    rooms = global_decision_room_service.rooms.list_rooms(status, decision_type, importance)
    return {"status": "SUCCESS", "count": len(rooms), "data": rooms}


@router.post("", status_code=status.HTTP_201_CREATED, summary="Create Decision Room")
async def create_decision_room(req: DecisionRoomCreateRequest) -> Dict[str, Any]:
    room = global_decision_room_service.rooms.create_room(
        title=req.title,
        question=req.question,
        owner_id=req.owner_id,
        decision_type=DecisionType(req.decision_type) if req.decision_type in DecisionType.__members__ else DecisionType.STRATEGIC,
        importance=DecisionImportance(req.importance) if req.importance in DecisionImportance.__members__ else DecisionImportance.MEDIUM,
        objective=req.objective,
        meta_info=req.meta_info,
    )
    return {"status": "SUCCESS", "data": room}


@router.get("/{id}", summary="Get Full Decision Room Workspace Overview")
async def get_decision_room_overview(id: str) -> Dict[str, Any]:
    try:
        overview = global_decision_room_service.get_room_overview(id)
        return {"status": "SUCCESS", "data": overview}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/{id}/transition", summary="Transition Decision Room Status")
async def transition_room_status(id: str, req: DecisionStatusTransitionRequest) -> Dict[str, Any]:
    try:
        room = global_decision_room_service.rooms.transition_status(
            room_id=id,
            target_status=DecisionStatus(req.target_status) if req.target_status in DecisionStatus.__members__ else DecisionStatus.OPEN,
            actor_id=req.actor_id,
            notes=req.notes,
        )
        return {"status": "SUCCESS", "data": room}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{id}/decide", summary="Record Final Human Executive Decision")
async def record_decision(id: str, req: RecordDecisionRequest) -> Dict[str, Any]:
    try:
        room = global_decision_room_service.rooms.record_decision(
            room_id=id,
            selected_option_id=req.selected_option_id,
            decision_summary=req.decision_summary,
            decided_by=req.decided_by,
        )
        return {"status": "SUCCESS", "data": room}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{id}/context", summary="Set Decision Room Context")
async def set_room_context(id: str, req: SetContextRequest) -> Dict[str, Any]:
    ctx = global_decision_room_service.context_evidence.set_context(
        room_id=id,
        background=req.background,
        current_state=req.current_state,
        constraints=req.constraints,
        entities_involved=req.entities_involved,
    )
    return {"status": "SUCCESS", "data": ctx}


@router.post("/{id}/evidence", summary="Add Evidence Item with Fact/Inference Category")
async def add_evidence_item(id: str, req: AddEvidenceRequest) -> Dict[str, Any]:
    ev_type = EvidenceType(req.evidence_type) if req.evidence_type in EvidenceType.__members__ else EvidenceType.DATABASE
    cat = StatementCategory(req.statement_category) if req.statement_category in StatementCategory.__members__ else StatementCategory.FACT
    item = global_decision_room_service.context_evidence.add_evidence(
        room_id=id,
        evidence_type=ev_type,
        source=req.source,
        claim=req.claim,
        statement_category=cat,
        authority=req.authority,
        confidence=req.confidence,
    )
    return {"status": "SUCCESS", "data": item}


@router.post("/{id}/assumptions", summary="Record Assumption Item")
async def add_assumption_item(id: str, req: AddAssumptionRequest) -> Dict[str, Any]:
    item = global_decision_room_service.assumptions_options.add_assumption(
        room_id=id,
        statement=req.statement,
        confidence=req.confidence,
        impact_if_false=req.impact_if_false,
    )
    return {"status": "SUCCESS", "data": item}


@router.post("/{id}/unknowns", summary="Record Known-Unknown Item")
async def add_unknown_item(id: str, req: AddUnknownRequest) -> Dict[str, Any]:
    item = global_decision_room_service.assumptions_options.add_unknown(
        room_id=id,
        question=req.question,
        impact=req.impact,
        resolution_path=req.resolution_path,
    )
    return {"status": "SUCCESS", "data": item}


@router.post("/{id}/options", summary="Create Candidate Option")
async def create_candidate_option(id: str, req: CreateOptionRequest) -> Dict[str, Any]:
    opt = global_decision_room_service.assumptions_options.create_option(
        room_id=id,
        name=req.name,
        description=req.description,
        benefits=req.benefits,
        costs=req.costs,
        risks=req.risks,
        uncertainty_level=req.uncertainty_level,
        reversibility=req.reversibility,
    )
    return {"status": "SUCCESS", "data": opt}


@router.post("/{id}/criteria", summary="Add Decision Evaluation Criterion")
async def add_decision_criterion(id: str, req: AddCriterionRequest) -> Dict[str, Any]:
    crit = global_decision_room_service.assumptions_options.add_criterion(
        room_id=id,
        name=req.name,
        weight=req.weight,
        criterion_type=req.criterion_type,
        description=req.description,
    )
    return {"status": "SUCCESS", "data": crit}


@router.post("/{id}/scores", summary="Score Option against Criterion")
async def score_option_criterion(id: str, req: ScoreOptionRequest) -> Dict[str, Any]:
    score = global_decision_room_service.assumptions_options.score_option(
        room_id=id,
        option_id=req.option_id,
        criterion_id=req.criterion_id,
        raw_score=req.raw_score,
        justification=req.justification,
        scored_by=req.scored_by,
    )
    return {"status": "SUCCESS", "data": score}


@router.post("/{id}/analyses", summary="Submit Specialist Domain Analysis")
async def submit_specialist_analysis(id: str, req: SubmitAnalysisRequest) -> Dict[str, Any]:
    role = SpecialistRole(req.specialist_role) if req.specialist_role in SpecialistRole.__members__ else SpecialistRole.STRATEGY
    anls = global_decision_room_service.specialists.submit_specialist_analysis(
        room_id=id,
        specialist_role=role,
        summary=req.summary,
        recommendations=req.recommendations,
        key_findings=req.key_findings,
        confidence=req.confidence,
        facts=req.facts,
        inferences=req.inferences,
    )
    return {"status": "SUCCESS", "data": anls}


@router.post("/{id}/adversarial-reviews", summary="Submit Adversarial Review Critique")
async def submit_adversarial_review(id: str, req: SubmitAdversarialReviewRequest) -> Dict[str, Any]:
    rev = global_decision_room_service.specialists.submit_adversarial_review(
        room_id=id,
        critique_summary=req.critique_summary,
        reviewer_role=req.reviewer_role,
        weak_assumptions=req.weak_assumptions,
        unintended_consequences=req.unintended_consequences,
        hidden_costs=req.hidden_costs,
    )
    return {"status": "SUCCESS", "data": rev}


@router.post("/{id}/comments", summary="Post Discussion Comment or Annotation")
async def post_discussion_comment(id: str, req: AddCommentRequest) -> Dict[str, Any]:
    comment = global_decision_room_service.discussions_approvals.add_comment(
        room_id=id,
        author_id=req.author_id,
        content=req.content,
        author_role=req.author_role,
        is_human=req.is_human,
        parent_id=req.parent_id,
        mentions=req.mentions,
        annotations=req.annotations,
    )
    return {"status": "SUCCESS", "data": comment}


@router.post("/{id}/approvals", summary="Record Approval Step Decision")
async def record_approval_action(id: str, req: RecordApprovalRequest) -> Dict[str, Any]:
    try:
        status_enum = ApprovalStatus(req.status) if req.status in ApprovalStatus.__members__ else ApprovalStatus.APPROVED
        appr = global_decision_room_service.discussions_approvals.record_approval_action(
            room_id=id,
            step_id=req.step_id,
            approver_id=req.approver_id,
            status=status_enum,
            notes=req.notes,
        )
        return {"status": "SUCCESS", "data": appr}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{id}/actions", summary="Create Action Item from Decision")
async def create_action_item(id: str, req: CreateActionRequest) -> Dict[str, Any]:
    action = global_decision_room_service.discussions_approvals.create_action_item(
        room_id=id,
        title=req.title,
        target_system=req.target_system,
        description=req.description,
        target_payload=req.target_payload,
        assigned_to=req.assigned_to,
    )
    return {"status": "SUCCESS", "data": action}


@router.post("/{id}/outcomes", summary="Record Measured Post-Execution Outcome")
async def record_decision_outcome(id: str, req: RecordOutcomeRequest) -> Dict[str, Any]:
    outcome = global_decision_room_service.outcomes_journal.record_outcome(
        room_id=id,
        metric_name=req.metric_name,
        expected_value=req.expected_value,
        actual_value=req.actual_value,
    )
    return {"status": "SUCCESS", "data": outcome}


@router.post("/{id}/post-reviews", summary="Submit Post-Decision Retrospective Review")
async def submit_post_decision_review(id: str, req: SubmitPostReviewRequest) -> Dict[str, Any]:
    prev = global_decision_room_service.outcomes_journal.submit_post_review(
        room_id=id,
        reviewed_by=req.reviewed_by,
        outcome_rating=req.outcome_rating,
        prediction_error=req.prediction_error,
        assumption_error=req.assumption_error,
        execution_error=req.execution_error,
        lessons_learned=req.lessons_learned,
    )
    return {"status": "SUCCESS", "data": prev}


@router.get("/templates/catalog", summary="List Pre-Configured Decision Room Templates")
async def list_decision_templates() -> Dict[str, Any]:
    templates = global_decision_room_service.outcomes_journal.list_templates()
    return {"status": "SUCCESS", "count": len(templates), "data": templates}


@router.post("/copilot/query", summary="Query Collaboration Copilot")
async def query_collaboration_copilot(req: CopilotQueryRequest) -> Dict[str, Any]:
    try:
        res = global_decision_room_service.ask_copilot(req.room_id, req.query)
        return {"status": "SUCCESS", "data": res}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/lib/api/decision_rooms.ts
/**
 * TypeScript API Client for Phase 53 — Unified Human-AI Collaboration, Decision Room & Augmented Intelligence Platform.
 */

export interface DecisionRoom {
  id: string;
  title: string;
  question: string;
  objective?: string;
  decision_type: string;
  importance: string;
  status: string;
  owner_id: string;
  selected_option_id?: string;
  decision_summary?: string;
  decided_at?: string;
  decided_by?: string;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface DecisionEvidence {
  id: string;
  evidence_type: string;
  source: string;
  claim: string;
  provenance?: string;
  authority: string;
  statement_category: string;
  confidence: number;
  freshness: string;
}

export interface DecisionOption {
  id: string;
  name: string;
  description: string;
  benefits: string[];
  costs: number;
  risks: string[];
  uncertainty_level: string;
  reversibility: string;
  composite_score: number;
  version: number;
}

export interface DecisionTradeoff {
  id: string;
  option_a_id: string;
  option_b_id: string;
  tradeoff_summary: string;
  gains_in_a: string[];
  sacrifices_in_a: string[];
}

export interface SpecialistAnalysis {
  id: string;
  specialist_role: string;
  worker_id?: string;
  summary: string;
  recommendations: string[];
  key_findings: string[];
  confidence: number;
  facts: string[];
  inferences: string[];
}

export interface AdversarialReview {
  id: string;
  reviewer_role: string;
  target_option_id?: string;
  critique_summary: string;
  weak_assumptions: string[];
  unintended_consequences: string[];
  hidden_costs: string[];
  data_gaps: string[];
}

export interface DecisionDisagreement {
  id: string;
  topic: string;
  disagreement_category: string;
  party_a: string;
  view_a: string;
  party_b: string;
  view_b: string;
  status: string;
}

export interface DecisionApproval {
  id: string;
  step_name: string;
  required_role: string;
  approver_id?: string;
  status: string;
  decision_notes?: string;
  reviewed_at?: string;
}

export interface DecisionRoomOverview {
  room: DecisionRoom;
  context?: any;
  evidence: DecisionEvidence[];
  assumptions: any[];
  unknowns: any[];
  options: DecisionOption[];
  criteria: any[];
  tradeoffs: DecisionTradeoff[];
  scenarios: any[];
  risks: any[];
  analyses: SpecialistAnalysis[];
  reviews: AdversarialReview[];
  disagreements: DecisionDisagreement[];
  consensus: {
    consensus_pct: number;
    disagreement_pct: number;
    total_specialist_analyses: number;
    total_disagreements: number;
    primary_disagreement: string;
  };
  discussions: any[];
  approvals: DecisionApproval[];
  actions: any[];
  outcomes: any[];
  post_reviews: any[];
}

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export const decisionRoomsApi = {
  async listRooms(status?: string, decisionType?: string): Promise<DecisionRoom[]> {
    const params = new URLSearchParams();
    if (status) params.append("status", status);
    if (decisionType) params.append("decision_type", decisionType);
    const res = await fetch(`${API_BASE}/api/v1/decision-rooms?${params.toString()}`);
    if (!res.ok) throw new Error("Failed to list decision rooms");
    const json = await res.json();
    return json.data;
  },

  async getRoomOverview(id: string): Promise<DecisionRoomOverview> {
    const res = await fetch(`${API_BASE}/api/v1/decision-rooms/${id}`);
    if (!res.ok) throw new Error(`Failed to load decision room ${id}`);
    const json = await res.json();
    return json.data;
  },

  async createRoom(payload: {
    title: string;
    question: string;
    owner_id?: string;
    decision_type?: string;
    importance?: string;
    objective?: string;
  }): Promise<DecisionRoom> {
    const res = await fetch(`${API_BASE}/api/v1/decision-rooms`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error("Failed to create decision room");
    const json = await res.json();
    return json.data;
  },

  async recordDecision(id: string, payload: { selected_option_id: string; decision_summary: string; decided_by?: string }): Promise<DecisionRoom> {
    const res = await fetch(`${API_BASE}/api/v1/decision-rooms/${id}/decide`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error("Failed to record decision");
    const json = await res.json();
    return json.data;
  },

  async recordApproval(id: string, payload: { step_id: string; approver_id: string; status?: string; notes?: string }): Promise<any> {
    const res = await fetch(`${API_BASE}/api/v1/decision-rooms/${id}/approvals`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error("Failed to record approval action");
    const json = await res.json();
    return json.data;
  },

  async queryCopilot(roomId: string, query: string): Promise<any> {
    const res = await fetch(`${API_BASE}/api/v1/decision-rooms/copilot/query`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ room_id: roomId, query }),
    });
    if (!res.ok) throw new Error("Failed to query collaboration copilot");
    const json = await res.json();
    return json.data;
  },
};
